### DISCUSSION TITLE
When I try to use JSBSim plugin in unreal engine 5, I get such error: LogJSBSim: Error: Trim Failed!!!

### DISCUSSION THREAD

**Discussion by 140695809** [JSBSim-Team/jsbsim#1132] [Q&A] [answered]
LogJSBSim: Error: Trim Failed!!!
![屏幕截图 2024-08-01 102741](https://github.com/user-attachments/assets/a13cf55f-5da0-47c2-9858-5089e7054c04)

**Comment by bcoconni** [comment]
By default, JSBSim throws an exception when it fails to trim the aircraft in a stable configuration.

You could either:
* Swallow the exception and ignore it.
* Skip the trim procedure altogether.
* Tweak the initial conditions until the trim procedure succeeds.

If you pick either of the 2 first options, you might see the aircraft spawn in an unstable condition (either below the ground or diving or turning, etc.)

**Reply by 140695809** [reply]
How to tweak the initial conditions?

**Reply by bcoconni** [reply]
Initial conditions are basically the position, attitude, velocity and flight control settings of the aircraft when it is spawned.
* Position is generally only relevant when the aircraft is trimmed on ground because it drives the ground heights under the aircraft carriage.
* Attitude is important because the Unreal plugin is set to trim the aircraft in level flight, so the attitude (pitch and roll) should be in a level flight i.e. pitch and roll should be close to zero. Heading is mostly irrelevant for the success of the trim algorithm except if there are some strong winds.
* Velocity is also important: too slow and the wings cannot generate enough lift, too fast and the engine cannot generate enough thrust.
* Control settings are the position of the elevator/ailerons/etc. which should be close to zero as well for the trim to succeed.
  * Of importance is that the engine should be running when trying to trim in flight. The trim algorithm will iterate on the engine throttle to tune the thrust to be exactly equal to the drag. Obviously if the engine if off, the trim will fail.

**Comment by seanmcleod** [comment]
In particular here is the default trim logic from the Unreal plugin.

https://github.com/JSBSim-Team/jsbsim/blob/328a0b1923279b3e0b44ad0f72da74123d64c14c/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L709-L739

Looking at the code and your output messages you can see that the Unreal plugin continues to run after a trim failure, just reporting the message you noted.

**Reply by 140695809** [reply]
I only received an error message, but I don't know the reason why it happened

**Comment by seanmcleod** [comment]
> I only received an error message, but I don't know the reason why it happened

As I showed with the code snippet above the Unreal plugin attempts to trim the aircraft, either on the ground or in the air. If the trim fails, then you will see this error message.

A trim solution is not guaranteed, see @bcoconni's reply above for some tips in terms of increasing the chances of finding a valid trim solution.

Now I noticed in your screenshot above that you're trying to use a helicopter model, the ah1s. I'm not sure how well if at all the trim code in JSBSim works with helicopters in flight. So, I'd suggest first trying it out on the ground, I'm more confident, I haven't tested it myself yet, that the ground trim would succeed.

**Reply by bcoconni** [reply]
> I noticed in your screenshot above that you're trying to use a helicopter model, the ah1s. I'm not sure how well if at all the trim code in JSBSim works with helicopters in flight.

I have not tested it either but I'm pretty sure it can only fail as, for an helicopter, the lift is produced by the rotor. So the throttle should be used to trim $\dot{W}$ i.e. the altitude but currently, the throttle is used to trim $\dot{U}$ i.e. the velocity.
https://github.com/JSBSim-Team/jsbsim/blob/b41fcddc505baf041a3082ac3be2312e392c7792/src/initialization/FGTrim.cpp#L808-L809

>  So, I'd suggest first trying it out on the ground, I'm more confident, I haven't tested it myself yet, that the ground trim would succeed.

I agree.

**Reply by bcoconni** [reply, marked as answer]
I guess we should open an issue to log the fact that the trim algorithm needs some adaptation for helicopters.

**Reply by 140695809** [reply]
Are you saying that the JSBSim plugin for Unreal Engine doesn't apply to helicopter flight control？

**Reply by bcoconni** [reply]
> I guess we should open an issue to log the fact that the trim algorithm needs some adaptation for helicopters.

Issue #1139 has been created to track this.

**Comment by seanmcleod** [comment]
No, just that trying to trim a helicopter in flight when you startup will always fail. So, for now you would always need to startup with the helicopter on the ground.

You will also need to confirm within the plugin that the you have the required input mappings, in particular do you have an input mapping for `fcs/collective-cmd-norm`? 

https://github.com/JSBSim-Team/jsbsim/blob/6cd02910972870292f055095336144bccd32711f/aircraft/ah1s/Systems/rotor_control.xml#L68-L72

**Reply by 140695809** [reply]
I try to startup with the helicopter on the ground. When the helicopter fly to the sky, I lost lose control of it because the yaw , pitch, roll change all the time and I can‘t stabilize the helicopter.

**Comment by seanmcleod** [comment]
Nobody said helicopters were easy to fly 😉
